File: src/main/java/world/ranking/RankingKeeper.java
package world.ranking;

import field.object.life.FieldMobTemplate;
import managers.MobManager;
import net.database.RankingAPI;

import java.util.*;
import java.util.stream.Collectors;

public class RankingKeeper {

    public final static RankingKeeper instance = new RankingKeeper();
    private boolean updating = false;

    // rankings
    private Map<String, PlayerRanking> characterData;
    private List<PlayerRanking> regular = new ArrayList<>();
    private List<PlayerRanking> hardcore = new ArrayList<>();
    private List<PlayerRanking> killCount = new ArrayList<>();
    private final Map<Integer, List<PlayerRanking>> mobKills = new LinkedHashMap<>();
    private final Map<Integer, List<PlayerRanking>> bossKills = new LinkedHashMap<>();
    // todo playtime
    // todo mastery
    // todo bosses
    // todo jq's

    // selection lists
    private final Set<Integer> mobs = new HashSet<>();
    private final Set<Integer> bosses = new HashSet<>();

    private RankingKeeper() {
    }

    public static RankingKeeper getInstance() {
        return instance;
    }

    public boolean isUpdating() {
        return updating;
    }

    public Map<String, PlayerRanking> getCharacterData() {
        return characterData;
    }

    public List<PlayerRanking> getRegular() {
        return regular;
    }

    public List<PlayerRanking> getHardcore() {
        return hardcore;
    }

    public List<PlayerRanking> getKillCount() {
        return killCount;
    }

    public Map<Integer, List<PlayerRanking>> getMobKills() {
        return mobKills;
    }

    public Map<Integer, List<PlayerRanking>> getBossKills() {
        return bossKills;
    }

    public Set<Integer> getBosses() {
        return bosses;
    }

    /**
     * For script use
     *
     * @param players List of players to check in
     * @param name    Name to check for
     * @return PlayerRanking object including the rank, or null if it doesn't exist
     */
    public AbstractMap.SimpleEntry<Integer, PlayerRanking> getRankByName(List<PlayerRanking> players, String name) {
        PlayerRanking player = players.stream()
                .filter(playerRanking -> playerRanking.getName().equals(name))
                .findFirst()
                .orElse(null);
        if (player == null) return null;
        Integer index = players.indexOf(player);
        return new AbstractMap.SimpleEntry<>(index, player);
    }

    /**
     * For script use
     *
     * @return A List of mob integers instead of a Set
     */
    public List<Integer> getMobs() {
        return new ArrayList<>(mobs);
    }

    public void updateAllRankings() {
        if (!updating) new Thread(new RankingUpdater()).start();
        else System.err.println("[RankingKeeper] Not done updating yet");
    }

    private class RankingUpdater implements Runnable {

        @Override
        public void run() {
            updating = true;
            characterData = RankingAPI.INSTANCE.getNonGMCharacters();
            updateRegularRanking();
            updateHardcoreRanking();
            updateKillCountRanking();
            updateMobKillsRanking();
            updateBossKillsRanking();
            updating = false;
            System.out.println("[RankingKeeper] All rankings have been updated!");
        }
    }

    private void updateRegularRanking() {
        List<PlayerRanking> players = new ArrayList<>(characterData.values());
        players.sort(Comparator.comparingInt(PlayerRanking::getLevel).reversed());
        regular = players;
    }

    private void updateHardcoreRanking() {
        hardcore = characterData.values().stream()
                .filter(PlayerRanking::isHardcore)
                .sorted(Comparator.comparingInt(PlayerRanking::getLevel).reversed())
                .collect(Collectors.toList());
    }

    private void updateKillCountRanking() {
        List<PlayerRanking> players = new ArrayList<>(characterData.values());
        players.sort(Comparator.comparingInt(PlayerRanking::getKillCount).reversed());
        killCount = players;
    }

    private void updateMobKillsRanking() {
        List<PlayerRanking> players = new ArrayList<>(characterData.values());
        players.forEach(player -> mobs.addAll(player.getMobKills()
                .keySet().stream()
                .filter(mob -> {
                    FieldMobTemplate template = MobManager.getMob(mob);
                    if (template == null) return false;
                    return !template.isBoss();
                }).collect(Collectors.toList())));
        mobs.forEach(mob -> updateMobKillsRanking(mobKills, mob));
    }

    private void updateBossKillsRanking() {
        List<PlayerRanking> players = new ArrayList<>(characterData.values());
        players.forEach(player -> bosses.addAll(player.getMobKills()
                .keySet().stream()
                .filter(mob -> {
                    FieldMobTemplate template = MobManager.getMob(mob);
                    if (template == null) return false;
                    return template.isBoss();
                }).collect(Collectors.toList())));
        bosses.forEach(mob -> updateMobKillsRanking(bossKills, mob));
    }

    private void updateMobKillsRanking(Map<Integer, List<PlayerRanking>> list, int id) {
        List<PlayerRanking> players = characterData.values().stream()
                .filter(playerRanking -> playerRanking.getMobKills().get(id) != null)
                .sorted((p1, p2) -> p2.getMobKills().get(id).compareTo(p1.getMobKills().get(id)))
                .collect(Collectors.toList());
        list.put(id, players); // override old data
    }
}

File: src/main/java/world/ranking/RankingKeeper.kt
package world.ranking

import client.mastery.MasteryType
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import managers.MobManager
import net.database.RankingAPI
import util.logging.LogType
import util.logging.Logger.log
import java.util.*
import java.util.stream.Collectors
import kotlin.collections.ArrayList
import kotlin.collections.HashSet
import kotlin.collections.LinkedHashMap
import kotlin.collections.set

object RankingKeeper {

    var isUpdating = false

    // rankings
    lateinit var characterData: Map<String, PlayerRanking>
    var regular: List<PlayerRanking> = ArrayList()
    var hardcore: List<PlayerRanking> = ArrayList()
    var killCount: List<PlayerRanking> = ArrayList()
    var totalDamage: List<PlayerRanking> = ArrayList()
    val mobKills: MutableMap<Int, List<PlayerRanking>> = LinkedHashMap()
    val bossKills: MutableMap<Int, List<PlayerRanking>> = LinkedHashMap()
    val masteries: MutableMap<MasteryType, List<PlayerRanking>> = LinkedHashMap()

    // todo playtime
    // todo jq's
    // selection lists
    private val mobs: MutableSet<Int> = HashSet()
    private val bosses: MutableSet<Int> = HashSet()
    val masteryTypes: MutableList<MasteryType> = ArrayList()

    /**
     * For script use
     *
     * @param players List of players to check in
     * @param name    Name to check for
     * @return PlayerRanking object including the rank, or null if it doesn't exist
     */
    fun getRankByName(players: List<PlayerRanking>, name: String): AbstractMap.SimpleEntry<Int, PlayerRanking>? {
        val player = players.stream()
                .filter { it.name == name }
                .findFirst()
                .orElse(null)
                ?: return null
        val index = players.indexOf(player)
        return AbstractMap.SimpleEntry(index, player)
    }

    /**
     * For script use
     *
     * @return A List of mob integers instead of a Set
     */
    fun getMobs(): List<Int> {
        return ArrayList(mobs)
    }

    fun getBosses(): List<Int> {
        return ArrayList(bosses)
    }

    fun updateAllRankings() {
        if (!isUpdating) {
            GlobalScope.launch {
                update()
            }
        } else log(LogType.MISC_CONSOLE, "Ranking not done updating yet", this)
    }

    private fun update() {
        isUpdating = true
        characterData = RankingAPI.nonGMCharacters
        updateRegularRanking()
        updateHardcoreRanking()
        updateKillCountRanking()
        updateTotalDamageRanking()
        updateMobKillsRanking()
        updateBossKillsRanking()
        updateMasteryRanking()
        isUpdating = false
        println("[RankingKeeper] All rankings have been updated!")
    }

    private fun updateRegularRanking() {
        val players = ArrayList(characterData.values)
        players.sortWith(Comparator.comparingInt { obj: PlayerRanking -> obj.level }.reversed())
        regular = players
    }

    private fun updateHardcoreRanking() {
        hardcore = characterData.values.stream()
                .filter { it.isHardcore }
                .sorted(Comparator.comparingInt { obj: PlayerRanking -> obj.level }.reversed())
                .collect(Collectors.toList())
    }

    private fun updateKillCountRanking() {
        val players = ArrayList(characterData.values)
        players.sortWith(Comparator.comparingInt { obj: PlayerRanking -> obj.killCount }.reversed())
        killCount = players
    }

    private fun updateTotalDamageRanking() {
        val players = ArrayList(characterData.values)
        players.sortWith(Comparator.comparingLong { obj: PlayerRanking -> obj.totalDamage }.reversed())
        totalDamage = players
    }

    private fun updateMobKillsRanking() {
        val players = ArrayList(characterData.values)
        players.forEach { player: PlayerRanking ->
            mobs.addAll(player.mobKills
                    .keys.stream()
                    .filter { !MobManager.getMob(it).isBoss }.collect(Collectors.toList()))
        }
        mobs.forEach { updateMobKillsRanking(mobKills, it) }
    }

    private fun updateBossKillsRanking() {
        val players = ArrayList(characterData.values)
        players.forEach { player: PlayerRanking ->
            bosses.addAll(player.mobKills
                    .keys.stream()
                    .filter { MobManager.getMob(it).isBoss }.collect(Collectors.toList()))
        }
        bosses.forEach { updateMobKillsRanking(bossKills, it) }
    }

    private fun updateMasteryRanking() {
        masteryTypes.clear()
        MasteryType.values().forEach { type ->
            masteryTypes.add(type)
            sortInnerList(masteries, type, { playerRank, index -> playerRank.masteries[index] })
        }
    }

    private fun updateMobKillsRanking(list: MutableMap<Int, List<PlayerRanking>>, id: Int) {
        sortInnerList(list, id, { playerRank, mobId -> playerRank.mobKills[mobId] })
    }

    private fun <T> sortInnerList(map: MutableMap<T, List<PlayerRanking>>, mapKey: T, valueGetter: (PlayerRanking, T) -> (Int?)) {
        map[mapKey] = characterData.values
                .filter { valueGetter.invoke(it, mapKey) != null }
                .sortedByDescending { valueGetter.invoke(it, mapKey) }
    }
}